types/helpers.d.ts

```diff
@@ -1,69 +1,100 @@
-type Computed = () => any;
-type MutationMethod = (...args: any[]) => void;
-type ActionMethod = (...args: any[]) => Promise<any>;
-type CustomVue = Vue & Dictionary<any>
-
-interface Mapper<R> {
-  (map: string[]): Dictionary<R>;
-  (map: Dictionary<string>): Dictionary<R>;
+/**
+ * Utility types to declare helper types
+ */
+type Computed<R> = () => R;
+type Method<R> = (...args: any[]) => R;
+type MutationMethod<P> = (payload: P) => void;
+type ActionMethod<P> = (payload: P) => Promise<any>;
+type CustomVue = Vue & Record<string, any>;
+
+interface BaseType { [key: string]: any }
+
+interface BaseStateMap<State, Getters> {
+  [key: string]: (this: CustomVue, state: State, getters: Getters) => any;
+}
+
+interface BaseMethodMap<F> {
+  [key: string]: (this: CustomVue, fn: F, ...args: any[]) => any;
+}
+
+/**
+ * mapGetters
+ */
+interface MapGetters<Getters> {
+  <G extends Getters = Getters, Key extends keyof G = keyof G>(map: Key[]): { [K in Key]: Computed<G[K]> };
+  <G extends Getters = Getters, Map extends Record<string, keyof G> = Record<string, keyof G>>(map: Map): { [K in keyof Map]: Computed<G[Map[K]]> };
+}
+
+interface RootMapGetters<Getters> extends MapGetters<Getters> {
+  <G extends Getters = Getters, Key extends keyof G = keyof G>(namespace: string, map: Key[]): { [K in Key]: Computed<G[K]> };
+  <G extends Getters = Getters, Map extends Record<string, keyof G> = Record<string, keyof G>>(namespace: string, map: Map): { [K in keyof Map]: Computed<G[Map[K]]> };
+}
+
+/**
+ * mapState
+ */
+interface MapState<State, Getters> {
+  <S extends State = State, Key extends keyof S = keyof S>(map: Key[]): { [K in Key]: Computed<S[K]> };
+  <S extends State = State, Map extends Record<string, keyof S> = Record<string, keyof S>>(map: Map): { [K in keyof Map]: Computed<S[Map[K]]> };
+  <S extends State = State, G extends Getters = Getters, Map extends BaseStateMap<S, G> = BaseStateMap<S, G>>(map: Map): { [K in keyof Map]: Computed<any> };
 }
 
-interface MapperWithNamespace<R> {
-  (namespace: string, map: string[]): Dictionary<R>;
-  (namespace: string, map: Dictionary<string>): Dictionary<R>;
+interface RootMapState<State, Getters> extends MapState<State, Getters> {
+  <S extends State = State, Key extends keyof S = keyof S>(namespace: string, map: Key[]): { [K in Key]: Computed<S[K]> };
+  <S extends State = State, Map extends Record<string, keyof S> = Record<string, keyof S>>(namespace: string, map: Map): { [K in keyof Map]: Computed<S[Map[K]]> };
+  <S extends State = State, G extends Getters = Getters, Map extends BaseStateMap<S, G> = BaseStateMap<S, G>>(namespace: string, map: Map): { [K in keyof Map]: Computed<any> };
 }
 
-interface FunctionMapper<F, R> {
-  (map: Dictionary<(this: CustomVue, fn: F, ...args: any[]) => any>): Dictionary<R>;
+/**
+ * mapMutations
+ */
+interface MapMutations<Mutations> {
+  <M extends Mutations = Mutations, Key extends keyof M = keyof M>(map: Key[]): { [K in Key]: MutationMethod<M[K]> };
```

Comment:
There is a problem with returning `MutationMethod` here.
`MutationMethod` does not have `...args: any[]` parameters anymore.
This makes all mutations take one argument, same goes for actions.

Reply:
It's not actually a problem because mutations and actions only accept one parameter. The returned methods will never process 2nd or latter parameters.

Reply:
![image](https://user-images.githubusercontent.com/2339406/34715364-0b4dbdee-f535-11e7-8395-ded27aa4c461.png)

Reply:
You can pass multiple parameters (or none) to mapped actions

Reply:
Ah, I see. It should be an optional argument in default to follow the existing behavior.

But I think if we annotate types via `createNamespacedHelper` explicitly, it should not be an optional. Because it easily breaks apps if the user adds a payload for actions/mutations with no payload and the type checker cannot detect that.

If the users want to declare non-payload actions/mutations, they should explicitly pass a `null` or `undefined` value.

@HerringtonDarkholme What do you think?

Reply:
I agree that users should explicitly pass `null` or `undefined`.

Reply:
I agree, i believe `undefined` is probably best to signal that this payload is void. `null` could have context to an application where undefined can't be used on the store as it breaks the reactivity principles.